### PR TITLE
Add a Guard authenticator implementation

### LINKED ISSUE #144: Add a guard implementation
Hi,

I am currently integrating the bundle into a brand new symfony 4.4 project. I wish to protect a certain route with two authentication methods, being JWT Token and OAuth2 Token, depending on the requesting party.
For the JWT Token Auth I use package 'lexik/jwt-authentication-bundle`, which ships with an integrated Security Guard implementation that makes it very easy to add to a firewall section. Your package does not, but I think it would be a great enhancement.

### PR DESCRIPTION
**Pull request by ajgarlag**
Fix #144 

Most of the code borrowed from @yceruto gist at https://gist.github.com/yceruto/f1500b350065481568a3b2971fde8d0f

### CODE REVIEW

**Comment by yceruto** [comment]
Thanks for this PR! I will check it this weekend.

**Comment by yceruto** [comment]
Will be good if you change the PR status to make Travis run.

File: Security/Guard/Authenticator/OAuth2Authenticator.php
Comment (ajgarlag) [inline Security/Guard/Authenticator/OAuth2Authenticator.php:55]: @yceruto I'm not sure why you need a JsonResponse here. I think that the response type should depend on the request format? WDYT?

Comment (yceruto) [reply Security/Guard/Authenticator/OAuth2Authenticator.php:55]: Yes, it's only my end. For general purpose we have to take "the request format" into account.

In Symfony, we use the `ErrorHandler/ErrorRenderer/SerializerErrorRenderer` (introduced in 4.4) to return the right error response according to the request format. Depending on the bundle's min requirements, we could use it.

Comment (ajgarlag) [reply Security/Guard/Authenticator/OAuth2Authenticator.php:55]: I've tested it and is doable, but it introduces a new dependency (`symfony/serializer`) and makes the guard authentication behaviour different from the old firewall.

I think that guard authenticator should mimic the firewall behaviour. Maybe we should introduce `ErrorHandler/ErrorRenderer/SerializerErrorRenderer` in a different PR

**Comment by froozeify** [comment]
Hey,

Firstly I wanted to thanks you about your work on that PR  :-)
And if you need some help on it I can offer some assistance.

While reading what you did and comparing to the Authenticator from the package `lexik/jwt-authentication-bundle`.
I found something very useful in it, and wanted to know if it could be possible to add it also in that PR.
Or, if you prefer, once this one is merged, I can create my own PR implementing it ;-)

The cool stuff they implemented is [**Event** inside the authenticator](https://github.com/lexik/LexikJWTAuthenticationBundle/blob/master/Security/Guard/JWTTokenAuthenticator.php#L244)

What do you think about that ? Could that be done inside this PR, or do you prefer that I add it later when this one is closed ?

File: Security/Guard/Authenticator/OAuth2Authenticator.php
Comment (yceruto) [inline Security/Guard/Authenticator/OAuth2Authenticator.php:95]: Yes, as @froozeify said here https://github.com/trikoder/oauth2-bundle/pull/242#issuecomment-726839920 would be good to implement an extension point to change the response and do what is done here https://github.com/lexik/LexikJWTAuthenticationBundle/blob/2c3448975c5560f4a29ac6f3bfc1fb24673609c0/Security/Guard/JWTTokenAuthenticator.php#L173 (Returning a JsonResponse by default).

The same applies to the `start()` response.

What do you think?

Comment (froozeify) [reply Security/Guard/Authenticator/OAuth2Authenticator.php:95]: I have some more idea, maybe it will be better to wait for this to be merged, and I'll make a merge request after ?

Comment (yceruto) [reply Security/Guard/Authenticator/OAuth2Authenticator.php:95]: Feel free to send the alternative proposal :)

Comment (ajgarlag) [reply Security/Guard/Authenticator/OAuth2Authenticator.php:95]: There are 3 ways to define a security layer in Symfony: Firewall, Guard Authenticator, and HTTP Authenticator (introduced in SF 5.1).

Introducing extension points using Events it's a good thing, but I think it should be added to every security layer implementation, so I'd prefer to see a different PR with your ideas for the Firewall implementation to see how it works.

I think that this PR should mimic Firewall behavior, so if the Event is introduced before this PR gets merged, I'll add it here.

Comment (froozeify) [reply Security/Guard/Authenticator/OAuth2Authenticator.php:95]: That works for me :)   
I'll try to create the PR this week, nice catch for the Firewall, it should totally have the Events too


@yceruto The other alternative (I will probably add it to a separated PR, didn't check but maybe somebody already asked for it) is to add a setting to configure a Response formatter.  
Currently, when an error occurred the response return is not formatted in any type. Just the message in the response.   
For example for my use case, it will be better that the response is in a JSON format (like lexik/jwt do, so my client can catch the error message).   
I was thinking of a setting where we could define a custom Response format (like `none`, `json` or a user custom response class).

For the moment it's just an idea, but a not a good one, like to know some opinion about it

**Comment by samzijlmans** [comment]
@ajgarlag Is there a eta for this merge request? Is there anything you need help with, I would be glad to help

**Comment by ajgarlag** [comment]
@samzijlmans Thanks for your interest. I'm waiting for maintainers review.

**Review by X-Coder264 — APPROVED** [review]

**Comment by X-Coder264** [comment]
@ajgarlag @yceruto Thank you for the effort that you've put into this PR.

@froozeify We are interested in adding extension points to the new Guard authenticator and to the "old" system so please feel free to send a PR now that this has been merged.

**Comment by froozeify** [comment]
No problem, I'm currently working on it. And I hope tomorrow a big part will be done :-)

For the moment I didn't create here the PR since it's still in a WIP state, but maybe I should create the PR here ? (first time working on a big PR) I'll take any advice about it :-)

**Pull request merged**
